### ISSUE TITLE
React warning in acf-pro-blocks.js at ACFInnerBlocks if <InnerBlocks /> is located at the top level in the render template

### ISSUE THREAD

**Issue by CreativeDive** [AdvancedCustomFields/acf#721]
Hey @lgladdy,

I noticed this warning in the browser console related to the `acf-pro-blocks.js`.

<img width="1129" alt="Bildschirmfoto 2022-09-29 um 09 09 52" src="https://user-images.githubusercontent.com/31563167/192963866-c64cf294-00ac-487a-9296-37565374e262.png">

Here the full log:

```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
    at ACFInnerBlocks (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:598:7)
    at BlockPreview (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:1275:3)
    at div
    at _BlockBody (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:856:7)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at <anonymous> (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18745)
    at BlockEdit (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:763:7)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40505
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:34158
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:36248
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:38432
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:1910
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:6467
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:9862
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:12656
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:1:4393
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24566
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at <anonymous> (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18745)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-library/index.min.js?ver=c1ac02bf1649c7c2ac55:105:17263
    at http://localhost:8888/wphave_dev/wp-content/plugins/woocommerce/packages/woocommerce-blocks/build/cart.js?ver=eca5dddfd4ad3527250558e061f74923:37:563
    at o (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:69:11589)
    at Yr (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40962)
    at ri (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:417)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:52834
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:34845
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:22517
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:72:6363
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:77:5875
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:2246
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:6804
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:10071
    at e (http://localhost:8888/wphave_dev/wp-content/plugins/woocommerce/packages/woocommerce-blocks/build/wc-blocks.js?ver=b87861e9b6a5f73cc203d7c447078ff3:1:4568)
    at WrappedBlockEdit (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:691:7)
    at o (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:69:11589)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18498
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24566
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at d.Component.shouldComponentUpdate (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18655)
    at Uh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:11842)
    at Wh
    at $r (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40276)
    at Rh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:7384)
    at div
    at ACFInnerBlocks (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:598:7)
    at div
    at BlockPreview (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:1275:3)
    at div
    at _BlockBody (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:856:7)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at <anonymous> (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18745)
    at BlockEdit (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:763:7)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40505
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:34158
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:36248
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:38432
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:1910
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:6467
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:9862
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:12656
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:1:4393
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24566
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at <anonymous> (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18745)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-library/index.min.js?ver=c1ac02bf1649c7c2ac55:105:17263
    at http://localhost:8888/wphave_dev/wp-content/plugins/woocommerce/packages/woocommerce-blocks/build/cart.js?ver=eca5dddfd4ad3527250558e061f74923:37:563
    at o (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:69:11589)
    at Yr (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40962)
    at ri (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:417)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:52834
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:34845
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:22517
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:72:6363
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:77:5875
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:2246
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:6804
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:79:10071
    at e (http://localhost:8888/wphave_dev/wp-content/plugins/woocommerce/packages/woocommerce-blocks/build/wc-blocks.js?ver=b87861e9b6a5f73cc203d7c447078ff3:1:4568)
    at WrappedBlockEdit (http://localhost:8888/wphave_dev/wp-content/plugins/wphave/inc/acf/plugin/acf/assets/build/js/pro/acf-pro-blocks.js?ver=6.0.1:691:7)
    at o (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:69:11589)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18498
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24566
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/data/index.min.js?ver=967e2297dd5da2ca3996:1:24046
    at d.Component.shouldComponentUpdate (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/compose/index.min.js?ver=f70191f4dafd04120d4b:1:18655)
    at Uh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:11842)
    at Wh
    at div
    at Hh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:10617)
    at Sh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:3609)
    at Gh (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:70:11718)
    at CC (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:114:13811)
    at div
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:69316
    at Kt (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:19:1678)
    at div
    at t (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:110113)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:110894
    at div
    at t (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:110113)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:110894
    at div
    at aC (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:114:5657)
    at Yt (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:19:2235)
    at div
    at div
    at div
    at div
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:1:17971
    at bo (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:48:61)
    at Wo (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/editor/index.min.js?ver=36499e591c1ac5620dba:3:6893)
    at Po (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:7:143)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:59231
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:12:57260
    at WithRegistryProvider(Component)
    at $r (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/block-editor/index.min.js?ver=b2da143bf2d4dc7bc46a:9:40276)
    at un (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/core-data/index.min.js?ver=bb67bb3593924b9ec1bb:1:45878)
    at un (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/core-data/index.min.js?ver=bb67bb3593924b9ec1bb:1:45878)
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/editor/index.min.js?ver=36499e591c1ac5620dba:53:1612
    at http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/editor/index.min.js?ver=36499e591c1ac5620dba:53:777
    at WithRegistryProvider(Component)
    at pd (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:188344)
    at md (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:189192)
    at vd (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/components/index.min.js?ver=3a34d2c8c290154a04d2:1:191037)
    at div
    at x (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/keyboard-shortcuts/index.min.js?ver=9a77e75c0efd974c1503:1:3978)
    at vo (http://localhost:8888/wphave_dev/wp-content/plugins/gutenberg/build/edit-post/index.min.js?ver=2068cdf45bb5d37cef81:52:821)
```

Is this reproducible for you or do you need more information?

**Comment by CreativeDive** [comment]
@lgladdy Any idea why this is happening? Could you please help solve this? This issue seems to occur in some cases when using inner blocks. Is the console output helpful for you?

**Comment by lgladdy** [comment]
Is <InnerBlocks /> the first thing inside your block template?

**Comment by CreativeDive** [comment]
@lgladdy thanks for response here. Why you asked this just I've this this out:

Initial case:

```
<button>...</button>
<InnerBlocks />
```

So `<InnerBlocks />` is not the first, but on level 1!

Second case:

```
<button>...</button>
<div>
    <InnerBlocks />
</div>
```

Wrapped `<InnerBlocks />` inside a div container and the console error is gone. In this case I don't need an extra enclosing div. Is there a way to avoid this error without an extra div?

**Comment by lgladdy** [comment]
Okay cool. Yeh, this is a tricky one. I imagine if your `<InnerBlocks />` is the first item in your template at the top level you'll get the same crash on preview that a `<!-- html comment -->` causes. This is because in order to remove all the markup, we have to use React refs can only do that if we can attach a ref to an item.

Whilst we could just skip adding ref's to InnerBlocks, if the ref doesn't exist we'll cause a fatal. Ref's are required for us to be able to fire `render_block_preview` including the block.

Maybe the answer here is that we prevent the errors, and add a console warning when we know the template won't work with `render_block_preview` so folks will know how to solve the issue.

**Comment by CreativeDive** [comment]
@lgladdy thanks for the explanation here. To solve this issue, do you need to wrap the `<InnerBlocks />` tag? Or what do you think is the best practice here? Or can we leave it at level 1 and have a less drastic console warning going forward?

**Comment by lgladdy** [comment]
This is resolved in ACF 6.0.4

**Comment by CreativeDive** [comment]
@lgladdy perfect, thank you. Much appreciated that my console is now free of red alerts :-)
